content/en/blog/_posts/2025-XX-XX-mutable-pv-affinity-alpha/index.md

```diff
@@ -0,0 +1,132 @@
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/zone
+        values:
+        - us-east1-b
+```
+to:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/region
+        values:
+        - us-east1
+```
+
+As another example, providers sometimes offer new generations of disks.
+New disks cannot always be attached to older nodes in the cluster.
+While this accessibility can also be expressed through PV node affinity and ensures the Pods can be scheduled to the right nodes,
+this can also prevent online disk upgrade.
+In this case, you may want to change the PV node affinity from:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: provider.com/disktype.gen1
+        values:
+        - available
+```
+to:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: provider.com/disktype.gen2
+        values:
+        - available
+```
+
+So, we are making it mutable now, a first step towards a more flexible online volume management.
+While it is a simple change that removes one validation from the API server,
+we still have a long way to go to integrate well with the Kubernetes ecosystem.
+
+## Try it out
+
+Currently, this feature is in alpha state.
+It is disabled by default, and may subject to change.
+To try it out, enable `MutablePVNodeAffinity` feature gate on APIServer, then you can edit PV spec.nodeAffinity field.
+Typically only administrators can edit PVs, please make sure you have the right RBAC permissions.
+
+Note that changing PV node affinity alone will not actually change the accessibility of the underlying volume.
+You must also update the underlying volume in the storage provider, and keep the node affinity in sync.
+
+## Race Condition between Editing and Scheduling
+
+There are only a few things out of Pod that can affects the scheduling decision. PV node affinity is one of them.
+It is fine to allow more nodes to access the volume by relaxing node affinity.
+But there is a race condition when you try to tighten node affinity:
+We don't know how scheduler will see our modified PV in its cache,
+so there is a small window where the scheduler may place a Pod on an old node that can no longer access the volume.
+
+One mitigation under discussion is to have the kubelet fail Pod startup if the PV’s node affinity is violated.
+This has not landed yet.
+So if you are trying this out now, please watch subsequent Pods that use the updated PV,
+and make sure they are scheduled onto nodes that can access the volume.
+
+## Future Integration with CSI (Container Storage Interface)
+
+Currently, it is up to the cluster administrator to modify both PV's node affinity and the underlying volume in the storage provider.
```

Comment:
What does a cluster admin need to do before making node affinity changes so that he/she won't run into problems?

Reply:
I think this should be explained in the "Try it out" section.